src/transformers/models/whisper/generation_whisper.py

```diff
@@ -1016,94 +1072,211 @@ def _set_language_and_task(language, task, is_multilingual, generation_config):
         elif hasattr(generation_config, "forced_decoder_ids") and generation_config.forced_decoder_ids is not None:
             forced_decoder_ids = generation_config.forced_decoder_ids
+        elif hasattr(config, "forced_decoder_ids") and config.forced_decoder_ids is not None:
+            forced_decoder_ids = config.forced_decoder_ids
         else:
-            forced_decoder_ids = kwargs.pop("forced_decoder_ids", None)
-
-        if task is not None or language is not None or (forced_decoder_ids is None and prompt_ids is not None):
-            forced_decoder_ids = []
-            if hasattr(generation_config, "language"):
-                if generation_config.language in generation_config.lang_to_id.keys():
-                    language_token = generation_config.language
-                elif generation_config.language in TO_LANGUAGE_CODE.keys():
-                    language_token = f"<|{TO_LANGUAGE_CODE[generation_config.language]}|>"
-                elif generation_config.language in TO_LANGUAGE_CODE.values():
-                    language_token = f"<|{generation_config.language}|>"
-                else:
-                    is_language_code = len(generation_config.language) == 2
-                    raise ValueError(
-                        f"Unsupported language: {generation_config.language}. Language should be one of:"
-                        f" {list(TO_LANGUAGE_CODE.values()) if is_language_code else list(TO_LANGUAGE_CODE.keys())}."
-                    )
-                if language_token not in generation_config.lang_to_id:
-                    raise ValueError(
-                        f"{language_token} is not supported by this specific model as it is not in the `generation_config.lang_to_id`."
-                        "(You should just add it to the generation config)"
-                    )
-                forced_decoder_ids.append((1, generation_config.lang_to_id[language_token]))
+            forced_decoder_ids = None
+
+        task = getattr(generation_config, "task", None)
+        language = getattr(generation_config, "language", None)
+
+        if forced_decoder_ids is not None and task is not None:
+            logger.info(
+                f"You have passed task={task}, but also have set `forced_decoder_ids` to {forced_decoder_ids} which creates a conflict. `forced_decoder_ids` will be ignored in favor of task={task}."
+            )
+            forced_decoder_ids = None
+        elif forced_decoder_ids is not None and language is not None:
+            logger.info(
+                f"You have passed language={language}, but also have set `forced_decoder_ids` to {forced_decoder_ids} which creates a conflict. `forced_decoder_ids` will be ignored in favor of language={language}."
+            )
+            forced_decoder_ids = None
+
+        init_tokens = [generation_config.decoder_start_token_id]
+        if forced_decoder_ids is not None and forced_decoder_ids[0][0] == 1:
```

Comment:
There could feasibly be an instance where the forced decoder ids start at position 2, e.g.
```python
from transformers import WhisperProcessor, WhisperForConditionalGeneration
from datasets import load_dataset, Audio

model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-tiny")
processor = WhisperProcessor.from_pretrained("openai/whisper-tiny")

dataset = load_dataset("mozilla-foundation/common_voice_16_1", "de", split="validation", streaming=True)
dataset = dataset.cast_column("audio", Audio(16_000))

sample = next(iter(dataset))
input_features = processor(sample["audio"]["array"], return_tensors="pt").input_features

# generate with default forced ids (en + transcribe)
pred_ids = model.generate(input_features)
pred_text = processor.batch_decode(pred_ids)
print(pred_text)

# generate with forced ids starting at idx 2
model.config.forced_decoder_ids = [[2, 50359]]
pred_ids = model.generate(input_features)
pred_text = processor.batch_decode(pred_ids)
print(pred_text)
```
**Print output:**
```python
['<|startoftranscript|><|en|><|transcribe|><|notimestamps|> His cabin is located on the street of Alpenminister at the moment.<|endoftext|>']
['<|startoftranscript|><|de|><|transcribe|><|notimestamps|> Seine Gebine rohnen heute auf dem Friedhof von Alpenminister bei Gemunden.<|endoftext|>']
```

=> the current logic is a bit brittle in that it would fail for such examples

Reply:
Hmm, but isn't that the expected behavior here actually? For the 2nd case, we don't define a language, but say the model should transcribe => so it detects German and transcribes.

When an "incorrectly" forced_decoder_ids is passed, a warning is thrown, but the forced_decoder_ids are still used => I improved the warning message though to make it clearer.

Reply:
See: https://github.com/huggingface/transformers/pull/28687/commits/524cd3c01fa2f13c15c81d3326c76a024518dc22